File: ios/HybridMyazaEmrtd.swift
import Foundation
import CoreNFC
import ImageIO
import UIKit
import CommonCrypto
import NitroModules

// ---------------------------------------------------------------------------
// HybridMyazaEmrtd — the iOS half of eMRTD chip reading, implementing the
// nitrogen-generated `HybridMyazaEmrtdSpec` from src/specs/MyazaEmrtd.nitro.ts.
// The Android half is HybridMyazaEmrtd.kt.
//
// Two responsibilities, deliberately no more:
//
//   • TRANSPORT. CoreNFC's ISO-7816 tag, wrapped so TypeScript can send an
//     APDU and get bytes back. Everything above — BAC, secure messaging, file
//     reads — is TypeScript, where it is tested against ICAO 9303's worked
//     examples in CI.
//   • PRIMITIVES. CommonCrypto's DES, AES and SHA. Apple's implementations are
//     audited and hardware-backed; a hand-written DES in JavaScript would be
//     both slower and a novel implementation on a path that authenticates a
//     passport.
//
// This file decides NOTHING about authenticity. It moves bytes.
// ---------------------------------------------------------------------------

final class HybridMyazaEmrtd: HybridMyazaEmrtdSpec {

  private var session: NFCTagReaderSession?
  private var tag: NFCISO7816Tag?
  private var delegate: SessionDelegate?

  // MARK: - Transport

  /// Whether this device can read chips.
  ///
  /// `NFCTagReaderSession.readingAvailable` has been observed returning FALSE on
  /// a device that reads chips perfectly well — across app restarts, a
  /// foreground transition and a reinstall — and on the very same run that then
  /// opened a session and read a passport's MRZ. A device that truly lacked the
  /// capability could not have got that far, so the false is the property
  /// lying, not the hardware answering.
  ///
  /// The likeliest reason is the thread: Nitro invokes this synchronously from
  /// the JS thread, and CoreNFC is UIKit-adjacent. Reading it on the main thread
  /// would be the direct fix, but a `DispatchQueue.main.sync` from the JS thread
  /// can deadlock whenever main is itself waiting on JS — a real possibility
  /// during startup, and far worse than a wrong boolean.
  ///
  /// So it exploits the one thing that IS certain: NFC hardware cannot vanish
  /// while the app is running. A single observed `true` is therefore
  /// authoritative for the rest of the process, and later false readings — which
  /// have only ever been wrong — cannot retract it. A device genuinely without
  /// NFC never observes a true and keeps reporting false, which is correct.
  private static var everAvailable = false

  func isAvailable() throws -> Bool {
    let now = NFCTagReaderSession.readingAvailable
    if now { Self.everAvailable = true }
    return now || Self.everAvailable
  }

  /// Decode an image format JS cannot render, returning base64 JPEG.
  ///
  /// Passport chips overwhelmingly store DG2 as JPEG 2000, which React Native
  /// cannot display — so the portrait the issuing government signed was being
  /// read, hash-verified against the SOD, and then thrown away unshown. iOS has
  /// a decoder: ImageIO registers `public.jpeg-2000`.
  ///
  /// Empty string on failure rather than throwing. The preview is a courtesy;
  /// the chip read's verdict does not depend on it.
  func decodeImage(dataBase64: String) throws -> Promise<String> {
    Promise.async {
      guard let data = Data(base64Encoded: dataBase64) else { return "" }
      // Off the main thread: a full-resolution portrait is a few hundred
      // kilobytes of wavelet data and decoding it should not stall a frame.
      return await withCheckedContinuation { (continuation: CheckedContinuation<String, Never>) in
        DispatchQueue.global(qos: .userInitiated).async {
          guard let source = CGImageSourceCreateWithData(data as CFData, nil),
                let cgImage = CGImageSourceCreateImageAtIndex(source, 0, nil),
                let jpeg = UIImage(cgImage: cgImage).jpegData(compressionQuality: 0.9)
          else {
            continuation.resume(returning: "")
            return
          }
          continuation.resume(returning: jpeg.base64EncodedString())
        }
      }
    }
  }

  func startSession(alertMessage: String) throws -> Promise<EmrtdTagInfo> {
    Promise.async {
      try await withCheckedThrowingContinuation { continuation in
        // The WHOLE session lifecycle runs on the MAIN thread. Nitro promises
        // execute on a background executor, and a session created and begun
        // there has been observed invalidating INSTANTLY with `Code=202
        // "Session invalidated unexpectedly"` — before the system sheet ever
        // appears — on the same phone that reads chips happily through a
        // main-thread session (the Flutter SDK's path; method channels deliver
        // on main). Consistent with `isAvailable` above: CoreNFC is
        // UIKit-adjacent and answers wrongly off the main thread. Unlike the
        // availability read, dispatching async here cannot deadlock — nothing
        // waits synchronously on the result.
        DispatchQueue.main.async {
          // A stale session from an earlier attempt must never overlap the new
          // one — two live sessions is itself an instant invalidation.
          self.session?.invalidate()
          self.session = nil
          self.tag = nil

          // The delegate is retained explicitly: NFCTagReaderSession holds its
          // delegate weakly, and a deallocated one silently ends the session
          // with no callback at all.
          let delegate = SessionDelegate { [weak self] result in
            switch result {
            case .success(let tag):
              self?.tag = tag
              continuation.resume(
                returning: EmrtdTagInfo(connected: true, uid: Self.identifier(of: tag))
              )
            case .failure(let error):
              continuation.resume(throwing: error)
            }
          }
          self.delegate = delegate

          guard
            let session = NFCTagReaderSession(
              pollingOption: .iso14443, delegate: delegate, queue: nil
            )
          else {
            continuation.resume(throwing: EmrtdError.unavailable)
            return
          }
          session.alertMessage = alertMessage
          self.session = session
          session.begin()
        }
      }
    }
  }

  func stopSession(message: String) throws -> Promise<Void> {
    Promise.async {
      await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
        // Same main-thread rule as startSession — teardown must not race a
        // begin() already queued on main.
        DispatchQueue.main.async {
          if !message.isEmpty {
            self.session?.alertMessage = message
          }
          self.session?.invalidate()
          self.session = nil
          self.tag = nil
          self.delegate = nil
          continuation.resume()
        }
      }
    }
  }

  func updateSessionMessage(message: String) throws -> Promise<Void> {
    Promise.async {
      await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
        // Same main-thread rule as the rest of the session lifecycle. The
        // system sheet covers the app during a read, so this line is the only
        // narration an iOS user gets — updated per stage from TypeScript.
        DispatchQueue.main.async {
          self.session?.alertMessage = message
          continuation.resume()
        }
      }
    }
  }

  func transceive(apduBase64: String) throws -> Promise<EmrtdApduResponse> {
    Promise.async {
      guard let tag = self.tag else { throw EmrtdError.notConnected }
      guard
        let raw = Data(base64Encoded: apduBase64),
        let apdu = NFCISO7816APDU(data: raw)
      else {
        throw EmrtdError.malformedApdu
      }

      let (data, sw1, sw2) = try await tag.sendCommand(apdu: apdu)
      // A non-success status word RESOLVES rather than throwing: 0x6CXX and
      // 0x63CF are answers the protocol layer has to interpret, not failures.
      return EmrtdApduResponse(
        data: data.base64EncodedString(),
        statusWord: Double(Int(sw1) << 8 | Int(sw2))
      )
    }
  }

  private static func identifier(of tag: NFCISO7816Tag) -> String {
    tag.identifier.base64EncodedString()
  }

  // MARK: - Primitives

  func sha1(dataBase64: String) throws -> String {
    let data = Self.decode(dataBase64)
    // The length is captured BEFORE the buffer is borrowed: reading
    // `data.count` inside `withUnsafeBytes` is an overlapping access.
    let length = CC_LONG(data.count)
    var digest = [UInt8](repeating: 0, count: Int(CC_SHA1_DIGEST_LENGTH))
    data.withUnsafeBytes { buffer in
      _ = CC_SHA1(buffer.baseAddress, length, &digest)
    }
    return Data(digest).base64EncodedString()
  }

  func sha256(dataBase64: String) throws -> String {
    let data = Self.decode(dataBase64)
    let length = CC_LONG(data.count)
    var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    data.withUnsafeBytes { buffer in
      _ = CC_SHA256(buffer.baseAddress, length, &digest)
    }
    return Data(digest).base64EncodedString()
  }

  func desEde2Cbc(
    keyBase64: String, dataBase64: String, ivBase64: String, encrypt: Bool
  ) throws -> String {
    // Two-key 3DES is K1|K2|K1. CommonCrypto wants the full 24 bytes, so the
    // expansion happens here rather than being assumed.
    let key16 = Self.decode(keyBase64)
    guard key16.count == 16 else { throw EmrtdError.badKeyLength }
    var key24 = Data(key16)
    key24.append(key16.prefix(8))

    return try Self.crypt(
      algorithm: CCAlgorithm(kCCAlgorithm3DES),
      key: key24,
      iv: Self.decode(ivBase64),
      data: Self.decode(dataBase64),
      blockSize: kCCBlockSize3DES,
      encrypt: encrypt
    )
  }

  func desBlock(keyBase64: String, blockBase64: String, encrypt: Bool) throws -> String {
    // ECB on a single block — the retail MAC's inner primitive. The chaining is
    // done in TypeScript, where the standard's construction is testable.
    try Self.crypt(
      algorithm: CCAlgorithm(kCCAlgorithmDES),
      key: Self.decode(keyBase64),
      iv: Data(),
      data: Self.decode(blockBase64),
      blockSize: kCCBlockSizeDES,
      encrypt: encrypt,
      options: CCOptions(kCCOptionECBMode)
    )
  }

  func aesCbc(
    keyBase64: String, dataBase64: String, ivBase64: String, encrypt: Bool
  ) throws -> String {
    try Self.crypt(
      algorithm: CCAlgorithm(kCCAlgorithmAES),
      key: Self.decode(keyBase64),
      iv: Self.decode(ivBase64),
      data: Self.decode(dataBase64),
      blockSize: kCCBlockSizeAES128,
      encrypt: encrypt
    )
  }

  func aesCmac(keyBase64: String, dataBase64: String) throws -> String {
    // CommonCrypto has no CMAC, so RFC 4493 is built on AES-ECB here. It is a
    // short, exactly-specified construction; the alternative is a third-party
    // dependency inside the passport path.
    let key = Self.decode(keyBase64)
    let message = Self.decode(dataBase64)
    let block = 16

    // Subkeys: encrypt a zero block, then shift-and-conditionally-XOR.
    let zeros = Data(repeating: 0, count: block)
    let l = Self.decode(
      try Self.crypt(
        algorithm: CCAlgorithm(kCCAlgorithmAES), key: key, iv: Data(), data: zeros,
        blockSize: kCCBlockSizeAES128, encrypt: true, options: CCOptions(kCCOptionECBMode)
      )
    )
    let k1 = Self.shiftLeftXor(l)
    let k2 = Self.shiftLeftXor(k1)

    // The final block is XORed with K1 when the message divides evenly, and
    // padded then XORed with K2 when it does not.
    var blocks = [Data]()
    var i = 0
    while i < message.count {
      blocks.append(message.subdata(in: i..<min(i + block, message.count)))
      i += block
    }
    if blocks.isEmpty { blocks = [Data()] }

    var last = blocks.removeLast()
    if last.count == block {
      last = Self.xor(last, k1)
    } else {
      var padded = last
      padded.append(0x80)
      padded.append(Data(repeating: 0, count: block - padded.count))
      last = Self.xor(padded, k2)
    }

    var chain = Data(repeating: 0, count: block)
    for b in blocks + [last] {
      let input = Self.xor(chain, b)
      chain = Self.decode(
        try Self.crypt(
          algorithm: CCAlgorithm(kCCAlgorithmAES), key: key, iv: Data(), data: input,
          blockSize: kCCBlockSizeAES128, encrypt: true, options: CCOptions(kCCOptionECBMode)
        )
      )
    }
    return chain.base64EncodedString()
  }

  func randomBytes(length: Double) throws -> String {
    var bytes = [UInt8](repeating: 0, count: Int(length))
    // SecRandomCopyBytes, not arc4random: the BAC nonce's unpredictability is
    // what stops the exchange being replayable.
    guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
      throw EmrtdError.randomFailed
    }
    return Data(bytes).base64EncodedString()
  }

  // MARK: - Helpers

  private static func decode(_ b64: String) -> Data {
    Data(base64Encoded: b64) ?? Data()
  }

  private static func xor(_ a: Data, _ b: Data) -> Data {
    Data(zip(a, b).map { $0 ^ $1 })
  }

  /// One-bit left shift, XORed with the RFC 4493 constant on overflow.
  private static func shiftLeftXor(_ input: Data) -> Data {
    var out = [UInt8](repeating: 0, count: input.count)
    var carry: UInt8 = 0
    for i in stride(from: input.count - 1, through: 0, by: -1) {
      let byte = input[i]
      out[i] = (byte << 1) | carry
      carry = (byte & 0x80) != 0 ? 1 : 0
    }
    if (input.first! & 0x80) != 0 { out[out.count - 1] ^= 0x87 }
    return Data(out)
  }

  private static func crypt(
    algorithm: CCAlgorithm,
    key: Data,
    iv: Data,
    data: Data,
    blockSize: Int,
    encrypt: Bool,
    options: CCOptions = 0
  ) throws -> String {
    // NO padding option: the protocol applies ISO 9797-1 method 2 itself, and
    // letting CommonCrypto add PKCS#7 on top would corrupt every message.
    var out = Data(count: data.count + blockSize)
    // Captured BEFORE the buffer is exclusively borrowed below — reading
    // `out.count` inside `withUnsafeMutableBytes` is an overlapping access.
    let capacity = out.count
    let keyLength = key.count
    let dataLength = data.count
    let hasIv = !iv.isEmpty
    var moved = 0
    let status = out.withUnsafeMutableBytes { outPtr in
      data.withUnsafeBytes { dataPtr in
        key.withUnsafeBytes { keyPtr in
          iv.withUnsafeBytes { ivPtr in
            CCCrypt(
              CCOperation(encrypt ? kCCEncrypt : kCCDecrypt),
              algorithm,
              options,
              keyPtr.baseAddress, keyLength,
              hasIv ? ivPtr.baseAddress : nil,
              dataPtr.baseAddress, dataLength,
              outPtr.baseAddress, capacity,
              &moved
            )
          }
        }
      }
    }
    guard status == kCCSuccess else { throw EmrtdError.cryptoFailed(Int(status)) }
    return out.prefix(moved).base64EncodedString()
  }
}

// MARK: - Session delegate

private final class SessionDelegate: NSObject, NFCTagReaderSessionDelegate {
  private let onResult: (Result<NFCISO7816Tag, Error>) -> Void
  private var settled = false

  init(onResult: @escaping (Result<NFCISO7816Tag, Error>) -> Void) {
    self.onResult = onResult
  }

  /// The continuation may be resumed exactly once; CoreNFC can deliver both an
  /// invalidation and a detection in edge cases, and resuming twice traps.
  private func settle(_ result: Result<NFCISO7816Tag, Error>) {
    guard !settled else { return }
    settled = true
    onResult(result)
  }

  func tagReaderSessionDidBecomeActive(_ session: NFCTagReaderSession) {}

  func tagReaderSession(_ session: NFCTagReaderSession, didInvalidateWithError error: Error) {
    settle(.failure(error))
  }

  func tagReaderSession(_ session: NFCTagReaderSession, didDetect tags: [NFCTag]) {
    guard let detected = tags.first, case let .iso7816(tag) = detected else {
      session.invalidate(errorMessage: "That is not a passport chip.")
      settle(.failure(EmrtdError.wrongTagType))
      return
    }
    session.connect(to: detected) { [weak self] error in
      if let error {
        self?.settle(.failure(error))
      } else {
        self?.settle(.success(tag))
      }
    }
  }
}

enum EmrtdError: Error {
  case unavailable
  case notConnected
  case malformedApdu
  case wrongTagType
  case badKeyLength
  case randomFailed
  case cryptoFailed(Int)
}

File: android/src/main/java/com/margelo/nitro/myazakyc/HybridMyazaEmrtd.kt
package com.margelo.nitro.myazakyc

import android.app.Activity
import android.nfc.NfcAdapter
import android.os.Bundle
import android.nfc.Tag
import android.nfc.tech.IsoDep
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import java.io.ByteArrayOutputStream
import com.facebook.react.bridge.ReactApplicationContext
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.security.MessageDigest
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

/** How long to wait for the user to present a document before giving up. */
private const val NFC_WAIT_SECONDS = 60L

/**
 * How long Android waits between "is the tag still there?" probes.
 *
 * Generous on purpose: the probes interleave with ISO-DEP traffic, and the
 * default (~125ms) is a known cause of chips dropping partway through a large
 * DG2 read. The read itself has its own timeout, so a tag genuinely removed is
 * still noticed.
 */
private const val PRESENCE_CHECK_DELAY_MS = 5_000

/**
 * Android half of eMRTD chip reading, implementing the nitrogen-generated
 * `HybridMyazaEmrtdSpec` from src/specs/MyazaEmrtd.nitro.ts. The iOS half is
 * HybridMyazaEmrtd.swift.
 *
 * Two responsibilities, deliberately no more:
 *
 *  - TRANSPORT. `IsoDep` reader-mode, wrapped so TypeScript can send an APDU and
 *    get bytes back. Everything above — BAC, secure messaging, file reads — is
 *    TypeScript, tested against ICAO 9303's worked examples in CI.
 *  - PRIMITIVES. javax.crypto's DES, AES and SHA. A hand-written DES in
 *    JavaScript would be slower and a novel implementation on a path that
 *    authenticates a passport.
 *
 * This file decides NOTHING about authenticity. It moves bytes.
 */
class HybridMyazaEmrtd : HybridMyazaEmrtdSpec() {

  // @Volatile because every entry point runs on an ARBITRARY pool thread
  // (Nitro's Promise.async dispatches to Dispatchers.Default) and the reader
  // callback lands on a binder thread. Without it the `isoDep = dep` write in
  // startSession has no happens-before edge to the read in transceive, and a
  // fresh connection can be invisible to the very next call — the iOS side
  // avoids this whole class by serializing on the main queue.
  @Volatile private var isoDep: IsoDep? = null
  @Volatile private var pending: CompletableFuture<Tag>? = null
  @Volatile private var adapter: NfcAdapter? = null

  private val context: ReactApplicationContext?
    get() = NitroModules.applicationContext

  // ── Transport ─────────────────────────────────────────────────────────────

  /**
   * Whether this device can read chips.
   *
   * DELIBERATELY not the "sticky true" the iOS side uses. There, a false is
   * always the platform lying about fixed hardware, so the first true is kept
   * forever. Here `isEnabled` reports a SETTING the user can turn off mid-flow,
   * so a false is real, actionable information and must be allowed to retract a
   * previous true — telling someone their chip can be read when they have just
   * disabled NFC would be worse than useless.
   *
   * Do not "harmonise" the two platforms: they are answering different
   * questions.
   */
  override fun isAvailable(): Boolean {
    val ctx = context ?: return false
    return NfcAdapter.getDefaultAdapter(ctx)?.isEnabled == true
  }

  /**
   * Decode an image format JS cannot render, returning base64 JPEG.
   *
   * Passport chips overwhelmingly store DG2 as JPEG 2000, which React Native
   * cannot display. BitmapFactory handles the formats the platform knows —
   * which does NOT include JPEG 2000 (an earlier comment here claimed it did;
   * it never has, which is why the portrait showed on iOS and never on
   * Android) — so JP2/J2K falls through to the bundled JJ2000 decoder.
   *
   * Empty string on failure rather than throwing: the preview is a courtesy and
   * the chip read's verdict does not depend on it.
   */
  override fun decodeImage(dataBase64: String): Promise<String> {
    return Promise.async {
      try {
        val bytes = Base64.decode(dataBase64, Base64.NO_WRAP)
        val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
          ?: Jp2Decoder.decode(bytes)
          ?: return@async ""
        val out = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out)
        Base64.encodeToString(out.toByteArray(), Base64.NO_WRAP)
      } catch (_: Throwable) {
        ""
      }
    }
  }

  override fun startSession(alertMessage: String): Promise<EmrtdTagInfo> {
    // `alertMessage` is iOS's system NFC sheet; Android shows nothing of its
    // own, so the SDK's own UI carries the instruction there.
    return Promise.async {
      val activity = context?.currentActivity
        ?: throw IllegalStateException("No foreground activity to attach the NFC reader to.")
      val nfc = NfcAdapter.getDefaultAdapter(activity)
        ?: throw IllegalStateException("This device has no NFC hardware.")

      val future = CompletableFuture<Tag>()
      pending = future
      adapter = nfc

      // Reader mode with NFC-A/B and the platform's own NDEF check SKIPPED: a
      // passport is not an NDEF tag, and letting the system probe for one adds
      // a round trip and can knock the chip off the field.
      //
      // The presence-check delay is the setting that decides whether a large
      // file survives. Android otherwise probes the tag every ~125ms to see if
      // it is still there, and those probes interleave with an ISO-DEP exchange
      // — long enough into a 30KB DG2 read, some chips drop the connection. The
      // library the Flutter SDK uses sets this for the same reason.
      val options = Bundle().apply {
        putInt(NfcAdapter.EXTRA_READER_PRESENCE_CHECK_DELAY, PRESENCE_CHECK_DELAY_MS)
      }
      nfc.enableReaderMode(
        activity,
        { tag -> future.complete(tag) },
        NfcAdapter.FLAG_READER_NFC_A or
          NfcAdapter.FLAG_READER_NFC_B or
          NfcAdapter.FLAG_READER_SKIP_NDEF_CHECK,
        options,
      )

      // BOUNDED. `get()` with no timeout waits forever for a tap that may never
      // come, and the caller would have no way out — the same unbounded-wait
      // trap the SDK avoids everywhere else. A user who does not present a
      // document gets a clean timeout instead of a frozen screen.
      val tag = try {
        future.get(NFC_WAIT_SECONDS, TimeUnit.SECONDS)
      } catch (e: TimeoutException) {
        throw IllegalStateException("No chip was detected. Try again and hold the phone still.")
      }
      val dep = IsoDep.get(tag)
        ?: throw IllegalStateException("That tag does not speak ISO-DEP.")
      // Passports are slow and DG2 is large; the default timeout aborts
      // mid-file on a chip that is working perfectly well.
      dep.timeout = 20_000
      dep.connect()
      isoDep = dep

      EmrtdTagInfo(connected = true, uid = Base64.encodeToString(tag.id, Base64.NO_WRAP))
    }
  }

  override fun stopSession(message: String): Promise<Unit> {
    return Promise.async {
      // Cancel a wait that is still in flight, so closing the sheet releases
      // the caller rather than leaving it blocked until the timeout.
      pending?.completeExceptionally(IllegalStateException("The scan was cancelled."))
      runCatching { isoDep?.close() }
      isoDep = null
      pending = null
      val activity = context?.currentActivity
      if (activity != null) runCatching { adapter?.disableReaderMode(activity) }
      adapter = null
    }
  }

  override fun updateSessionMessage(message: String): Promise<Unit> {
    // Android has no system NFC sheet to write to — the SDK's own screen
    // narrates the read there. Accepted and ignored so the JS side can call
    // one API on both platforms.
    return Promise.async { }
  }

  override fun transceive(apduBase64: String): Promise<EmrtdApduResponse> {
    return Promise.async {
      val dep = isoDep ?: throw IllegalStateException("No chip is connected.")
      val response = dep.transceive(Base64.decode(apduBase64, Base64.NO_WRAP))
      if (response.size < 2) throw IllegalStateException("The chip returned a truncated response.")

      // A non-success status word RESOLVES rather than throwing: 0x6CXX and
      // 0x63CF are answers the protocol layer must interpret, not failures.
      val sw1 = response[response.size - 2].toInt() and 0xFF
      val sw2 = response[response.size - 1].toInt() and 0xFF
      val data = response.copyOfRange(0, response.size - 2)

      EmrtdApduResponse(
        data = Base64.encodeToString(data, Base64.NO_WRAP),
        statusWord = ((sw1 shl 8) or sw2).toDouble(),
      )
    }
  }

  // ── Primitives ────────────────────────────────────────────────────────────

  override fun sha1(dataBase64: String): String = digest("SHA-1", dataBase64)

  override fun sha256(dataBase64: String): String = digest("SHA-256", dataBase64)

  private fun digest(algorithm: String, dataBase64: String): String =
    Base64.encodeToString(
      MessageDigest.getInstance(algorithm).digest(decode(dataBase64)),
      Base64.NO_WRAP,
    )

  override fun desEde2Cbc(
    keyBase64: String,
    dataBase64: String,
    ivBase64: String,
    encrypt: Boolean,
  ): String {
    // Two-key 3DES is K1|K2|K1. javax.crypto's DESede wants all 24 bytes, so
    // the expansion happens here rather than being assumed.
    val key16 = decode(keyBase64)
    require(key16.size == 16) { "3DES needs a 16-byte two-key value." }
    val key24 = ByteArray(24)
    System.arraycopy(key16, 0, key24, 0, 16)
    System.arraycopy(key16, 0, key24, 16, 8)

    val iv = decode(ivBase64)
    // NoPadding: the protocol applies ISO 9797-1 method 2 itself, and letting
    // the provider add PKCS#5 on top would corrupt every message.
    return crypt(
      "DESede/CBC/NoPadding",
      SecretKeySpec(key24, "DESede"),
      IvParameterSpec(if (iv.isEmpty()) ByteArray(8) else iv),
      decode(dataBase64),
      encrypt,
    )
  }

  override fun desBlock(keyBase64: String, blockBase64: String, encrypt: Boolean): String =
    // ECB on a single block — the retail MAC's inner primitive. The chaining is
    // done in TypeScript, where the standard's construction is testable.
    crypt(
      "DES/ECB/NoPadding",
      SecretKeySpec(decode(keyBase64), "DES"),
      null,
      decode(blockBase64),
      encrypt,
    )

  override fun aesCbc(
    keyBase64: String,
    dataBase64: String,
    ivBase64: String,
    encrypt: Boolean,
  ): String {
    val iv = decode(ivBase64)
    return crypt(
      "AES/CBC/NoPadding",
      SecretKeySpec(decode(keyBase64), "AES"),
      IvParameterSpec(if (iv.isEmpty()) ByteArray(16) else iv),
      decode(dataBase64),
      encrypt,
    )
  }

  override fun aesCmac(keyBase64: String, dataBase64: String): String {
    // No CMAC in the platform providers, so RFC 4493 is built on AES-ECB here.
    // A short, exactly-specified construction; the alternative is a
    // third-party dependency inside the passport path.
    val key = SecretKeySpec(decode(keyBase64), "AES")
    val message = decode(dataBase64)
    val block = 16

    val l = decode(crypt("AES/ECB/NoPadding", key, null, ByteArray(block), true))
    val k1 = shiftLeftXor(l)
    val k2 = shiftLeftXor(k1)

    val blocks = mutableListOf<ByteArray>()
    var i = 0
    while (i < message.size) {
      blocks.add(message.copyOfRange(i, minOf(i + block, message.size)))
      i += block
    }
    if (blocks.isEmpty()) blocks.add(ByteArray(0))

    // The final block is XORed with K1 when the message divides evenly, and
    // padded then XORed with K2 when it does not.
    var last = blocks.removeAt(blocks.size - 1)
    last = if (last.size == block) {
      xor(last, k1)
    } else {
      val padded = ByteArray(block)
      System.arraycopy(last, 0, padded, 0, last.size)
      padded[last.size] = 0x80.toByte()
      xor(padded, k2)
    }

    var chain = ByteArray(block)
    for (b in blocks + listOf(last)) {
      chain = decode(crypt("AES/ECB/NoPadding", key, null, xor(chain, b), true))
    }
    return Base64.encodeToString(chain, Base64.NO_WRAP)
  }

  override fun randomBytes(length: Double): String {
    val bytes = ByteArray(length.toInt())
    // SecureRandom, not Random: the BAC nonce's unpredictability is what stops
    // the exchange being replayable.
    SecureRandom().nextBytes(bytes)
    return Base64.encodeToString(bytes, Base64.NO_WRAP)
  }

  // ── Helpers ───────────────────────────────────────────────────────────────

  private fun decode(b64: String): ByteArray =
    if (b64.isEmpty()) ByteArray(0) else Base64.decode(b64, Base64.NO_WRAP)

  private fun xor(a: ByteArray, b: ByteArray): ByteArray =
    ByteArray(minOf(a.size, b.size)) { (a[it].toInt() xor b[it].toInt()).toByte() }

  /** One-bit left shift, XORed with the RFC 4493 constant on overflow. */
  private fun shiftLeftXor(input: ByteArray): ByteArray {
    val out = ByteArray(input.size)
    var carry = 0
    for (i in input.indices.reversed()) {
      val byte = input[i].toInt() and 0xFF
      out[i] = ((byte shl 1) or carry).toByte()
      carry = if (byte and 0x80 != 0) 1 else 0
    }
    if (input[0].toInt() and 0x80 != 0) {
      out[out.size - 1] = (out[out.size - 1].toInt() xor 0x87).toByte()
    }
    return out
  }

  private fun crypt(
    transformation: String,
    key: SecretKeySpec,
    iv: IvParameterSpec?,
    data: ByteArray,
    encrypt: Boolean,
  ): String {
    val cipher = Cipher.getInstance(transformation)
    val mode = if (encrypt) Cipher.ENCRYPT_MODE else Cipher.DECRYPT_MODE
    if (iv == null) cipher.init(mode, key) else cipher.init(mode, key, iv)
    return Base64.encodeToString(cipher.doFinal(data), Base64.NO_WRAP)
  }
}
